Aggregate projections renames

oskardudycz opened the pull request:
- [x] Renamed `SelfAggregate` registration to `Snapshot` to make the intention more explicit.  Added an explicit `SnapshotLifecycle` enum with `Live` and `Async` options (as there's no sense in performing a snapshot live).
- [x] Added LiveStreamAggregation for Code Generation needs. It registers the entity as a `Live` aggregation. 
- [x] Renamed `SingleStreamAggregation` into `SingleStreamProjection`, leaving the old name as a derived obsolete class.
- [x] Renamed `MultiStreamAggregation` into `MultiStreamProjection`, leaving the old name as a derived obsolete class.
- [x] Renamed `CustomAggregation` into `CustomProjection`
- [x] Renamed `ExperimentalMultiStreamAggregation` into `ExperimentalMultiStreamProjection`
- [x] Update docs

_**Note:** I'll need to revamp our Event Sourcing docs, but I'll do that in the dedicated PR. For now, I just adjusted the naming in the docs to be aligned with the recent convention._

oskardudycz commented:
@jeremydmiller, @mysticmind, the code changes are ready to be reviewed. I need to adjust (or in fact rewrite docs).

mysticmind reviewed (approved):
Looks good to me. The names are lot more clearer and well defined. Update the docs 👍

mysticmind commented:
In terms of docs, use both the terms Snapshot and single-stream projection referring to method `Snapshot`

oskardudycz commented:
@mysticmind, thank you for the review. Glad that you liked it!

I made one more change in https://github.com/JasperFx/marten/pull/2525/commits/8dd6781dc7fc2de9d57d15b6ff1adfe5fd094ecd:
- Added LiveStreamAggregation for Code Generation needs. It registers the entity as Live aggregation.
- Added an explicit SnapshotLifecycle enum with Live and Async values (as there's no sense in performing Live snapshot).

I will work on docs updates in the next few days.
